File: app/assets/javascripts/time_balancer.js
'use strict';

const moment = require('moment');

const daysBetween = function (startTime, endTime) {
  return endTime.diff(startTime, 'days');
};

const timeframes = function (time) {
  const endOfWeek = moment(time).endOf('week');
  let endOfMonth = moment(time).endOf('month');

  if (endOfMonth.isBefore(endOfWeek)) { endOfMonth = endOfWeek; }

  let endOfQuarter = moment(time).endOf('quarter');

  if (endOfQuarter.isBefore(endOfWeek)) { endOfQuarter = endOfWeek; }

  let endOfYear = moment(time).endOf('year');

  if (endOfYear.isBefore(endOfWeek)) { endOfYear = endOfWeek; }

  return {
    today: 1,
    week: daysBetween(time, endOfWeek),
    month: daysBetween(endOfWeek, endOfMonth),
    quarter: daysBetween(endOfMonth, endOfQuarter),
    year: daysBetween(endOfQuarter, endOfYear)
  };
};

const TimeBalancer = {
  baseBalances: function (time) {
    return timeframes(moment(time));
  }
};

module.exports = TimeBalancer;

File: app/poros/time_balancer.rb
# frozen_string_literal: true

# How many days' worth of productivity each timeframe is allowed to hold. Every
# timeframe starts where the previous one ended, and none of them may end before
# the end of this week.
class TimeBalancer
  def self.base_balances(time = Time.zone.now)
    new(time).()
  end

  def initialize(time)
    @time = time
  end

  def call
    {
      today: 1,
      week: days_between(time, end_of_week),
      month: days_between(end_of_week, end_of_month),
      quarter: days_between(end_of_month, end_of_quarter),
      year: days_between(end_of_quarter, end_of_year),
    }
  end

  private

  attr_reader :time

  def end_of_week
    @end_of_week ||= time.end_of_week(:sunday)
  end

  def end_of_month
    [time.end_of_month, end_of_week].max
  end

  def end_of_quarter
    [time.end_of_quarter, end_of_week].max
  end

  def end_of_year
    [time.end_of_year, end_of_week].max
  end

  def days_between(start_time, end_time)
    ((end_time - start_time) / 1.day).truncate
  end
end
